File: LCS/Java/lcs.java
import java.util.Scanner;

class lcs{
    public static void main(String [] args){
        Scanner tsr = new Scanner(System.in);

        System.out.println("Inserisci la prima stringa:");
        String first = tsr.nextLine();
        System.out.println("");

        System.out.println("Insersci la seconda stringa:");
        String second = tsr.nextLine();
        System.out.println("");
        
        int m = first.length();
        int n = second.length();
        char [] firstString = first.toCharArray();
        char [] secondString = second.toCharArray();

        System.out.println(reverse(rebuildTable(
            buildTable(m, n, firstString, secondString), 
                m, n, firstString, secondString, "")));
        tsr.close();

    }



    public static String rebuildTable(int table[][], int m, int n, 
            char first[], char second[], String LCS){
        if(m == 0 && n == 0){
            return "";
        }else{
            //cella a sinistra e sopra uguali 
            if(table[m][n] == table[m - 1][n] && table[m][n] == table[m][n - 1]){
                if(first[m - 1] == second[n - 1]){
                    LCS = first[m - 1] + 
                        rebuildTable(table, m - 1, n - 1, first, second, LCS);
                }else{
                    LCS = rebuildTable(table, m - 1, n, first, second, LCS);
                }
            }else{
                    if(first[m - 1] == second[n - 1]){
                        LCS = first[m - 1] + rebuildTable(table, m - 1, n - 1, first, 
                        second, LCS);
                    }else{
                        if(table[m - 1][n] >= table[m][n - 1]){
                            LCS = rebuildTable(table, m - 1, n, first, second, LCS);
                        }else{
                            LCS = rebuildTable(table, m, n - 1, first, second, LCS);
                        }
                    }
            }

        }
        return LCS;
    }


    public static int[][] buildTable(int m, int n, char firstString[], char secondString[]){
        int [][] C = new int[m + 1][n + 1];
        for(int i = 0; i < m + 1; i ++){
            for(int j = 0; j < n + 1; j++){
                if((i == 0 && j == 0) || (i == 0 && j != 0) || (i != 0 && j == 0)){
                    C[i][j] = 0;
                }else{
                    if(firstString[i - 1] == secondString[j - 1]){
                        C[i][j] = C[i-1][j-1] + 1;
                    }else{
                        if(C[i][j - 1] > C[i - 1][j]){
                            C[i][j] = C[i][j - 1];
                        }else{
                            C[i][j] = C[i - 1][j];
                        }
                        
                    }
                }
                
            }
        }
        return C;
    }

    public static String reverse(String string){
        char support[] = string.toCharArray();
        String newString = "";
        for(int i = support.length - 1; i >= 0; i--){
            newString += support[i];
        }
        return newString;
    }


}

File: LCS/Python/lcs.py
def printTable(table) -> None:
    for row in table:
        for element in row:
            print(element, end = ' ')
        print()

def buildTable(firstVector, secondVector) -> list:
    m = int(len(firstVector))
    n = int(len(secondVector))
    table = [[0 for x in range(n + 1)] for y in range(m + 1)]
    for i in range(m + 1):
        for j in range(n + 1):
            if (i == 0 and j == 0) or (i == 0 and j != 0) or (i != 0 and j == 0):
                table[i][j] = 0
            elif firstVector[i - 1] == secondVector[j - 1]:
                table[i][j] = table[i-1][j-1] + 1
            elif table[i][j - 1] > table[i - 1][j]:
                table[i][j] = table[i][j - 1]
            else:
                table[i][j] = table[i - 1][j]
    printTable(table)
    return table



def rebuiltTable(table, m, n, firstVector, secondVector, LCS) -> str:
    if m == 0 and n == 0:
        return ""
    else:
        if table[m][n] == table[m - 1][n] and table[m][n] == table[m][n - 1]:
            if firstVector[m - 1] == secondVector[n - 1]: 
                    LCS = firstVector[m - 1] + rebuiltTable(table, m - 1, n - 1, firstVector, secondVector, LCS)
            else:
                    LCS = rebuiltTable(table, m - 1, n, firstVector, secondVector, LCS)
        else:
            if firstVector[m - 1] == secondVector[n - 1]:
                        LCS = firstVector[m - 1] + rebuiltTable(table, m - 1, n - 1, firstVector, secondVector, LCS)
            else:
                if table[m - 1][n] >= table[m][n - 1]:
                    LCS = rebuiltTable(table, m - 1, n, firstVector, secondVector, LCS)
                else:
                    LCS = rebuiltTable(table, m, n - 1, firstVector, secondVector, LCS)

    return LCS     



def reverse(LCS) -> str:
    LCSList = list(LCS)
    string = ""
    max = len(LCSList) - 1
    for x in range(max, -1, -1):
        string = string + LCSList[x]
    
    return string
    



string1 = input("Inserisci stringa 1\n")
string2 = input("Inserisci stringa 2\n")

firstVector = list(string1)
secondVector = list(string2)



print("LCS: " + reverse(
        rebuiltTable(
            buildTable(firstVector, secondVector), len(firstVector), len(secondVector), firstVector, secondVector, "")))

